spackbot/pr_add_labels.py

```diff
@@ -0,0 +1,168 @@
+    "new-variant": {"patch": r"\+ +variant\("},
+    "resources": {"patch": r"\+ +resource\("},
+    #
+    # Functions
+    #
+    "external-packages": {"patch": r"[+-] +def determine_spec_details\("},
+    "libraries": {"patch": r"[+-] +def libs\("},
+    "headers": {"patch": r"[+-] +def headers\("},
+    "smoke-tests": {"patch": r"[+-] +def test\("},
+    #
+    # Core spack
+    #
+    "architecture": {
+        "filename": r"^lib/spack/spack/(architecture|operating_systems|platforms)"
+    },
+    "binary-packages": {"filename": r"^lib/spack/spack/binary_distribution"},
+    "build-environment": {"filename": r"^lib/spack/spack/build_environment"},
+    "build-systems": {"filename": r"^lib/spack/spack/build_systems"},
+    "new-command": {
+        "filename": r"^lib/spack/spack/cmd/[^/]+.py$",
+        "status": r"^added$",
+    },
+    "commands": {
+        "filename": r"^lib/spack/spack/cmd/[^/]+.py$",
+        "status": [r"^modified$", r"^modified$"],
+    },
+    "compilers": {"filename": r"^lib/spack/spack/compiler"},
+    "directives": {"filename": r"^lib/spack/spack/directives"},
+    "environments": {"filename": r"^lib/spack/spack/environment"},
+    "fetching": {"filename": r"^lib/spack/spack/(fetch|url|util/url|util/web)"},
+    "locking": {"filename": r"^lib/spack/(spack|llnl)/util/lock"},
+    "modules": {"filename": r"^lib/spack/spack/modules"},
+    "stage": {"filename": r"^lib/spack/spack/stage"},
+    "tests": {"filename": r"^lib/spack/spack/test"},
+    "utilities": {"filename": r"^lib/spack/spack/util"},
+    "versions": {"filename": r"^lib/spack/spack/version"},
+    #
+    # Documentation
+    #
+    "documentation": {"filename": r"^lib/spack/docs"},
+    #
+    # GitHub
+    #
+    "travis": {"filename": r"^\.travis"},
+    "actions": {"filename": r"^\.github/actions"},
+    "workflow": {"filename": r"^\.github/workflows"},
+    "git": {"filename": r"^\.gitignore"},
+    "flake8": {"filename": r"^\.flake8"},
+    "licenses": {"filename": r"^LICENSE"},
+    "gitlab": {"filename": r"^share/spack/gitlab"},
+    #
+    # Other
+    #
+    "defaults": {"filename": r"^etc/spack/defaults"},
+    "vendored-dependencies": {"filename": r"^lib/spack/external"},
+    "sbang": {"filename": r"sbang"},
+    "docker": {"filename": [r"[Dd]ockerfile$", r"^share/spack/docker"]},
+    "shell-support": {"filename": r"^share/spack/.*sh$"},
+}
+
+
+# compile all the regexes above, and ensure that all pattern dict values are lists
+for label, pattern_dict in label_patterns.items():
+    for attr in pattern_dict.keys():
+        patterns = pattern_dict[attr]
+        if not isinstance(patterns, list):
+            patterns = [patterns]
+        pattern_dict[attr] = [re.compile(s) for s in patterns]
+
+
+@router.register("pull_request", action="opened")
+@router.register("pull_request", action="synchronize")
+async def label_pull_requests(event, gh, *args, session, **kwargs):
+    """ Add labels to PRs based on which files were modified."""
```

Comment:
PEP-8 seems to suggest that docstrings shouldn't start with a space: https://www.python.org/dev/peps/pep-0008/#documentation-strings